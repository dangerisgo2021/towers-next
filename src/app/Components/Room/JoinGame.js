import React from "react";
import { Button, Row, Progress } from "antd";
import { get } from "lodash";
import { joinRoomClicked } from "../../../state/redux/room/actions";
import { useDispatch, useSelector } from "react-redux";
import { getProfile } from "../../../state/redux/auth/selectors/getProfile";
import { useRouter } from "next/router";

const defaultGameConfig = { minPlayers: 2 };
const defaultProfile = { name: "Anon ?????" };

const useGetRoomIdFromUrl = () => {
  const { query = {} } = useRouter();
  const { roomId } = query;
  return roomId;
};

const useJoinGameContainer = ({ gameConfig, players }) => {
  const dispatch = useDispatch();
  const roomId = useGetRoomIdFromUrl();
  const { id } = useSelector(getProfile) || defaultProfile;
  const joined = players.find((player) => get(player, "profile.id") === id);
  const currentPlayerCount = Array.isArray(players) ? players.length : 0;
  const { minPlayers, maxPlayers } = gameConfig;
  const percent = (currentPlayerCount / minPlayers) * 100;
  const isJoinDisabled = joined || currentPlayerCount >= maxPlayers;
  const isLeaveDisabled = !joined;

  return {
    percent,
    isJoinDisabled,
    isLeaveDisabled,
    joinRoomClicked: () => {
      dispatch(joinRoomClicked({ roomId }));
    },
  };
};
export const JoinGame = ({ gameConfig = defaultGameConfig, players }) => {
  const {
    joinRoomClicked,
    percent,
    isJoinDisabled,
    isLeaveDisabled,
  } = useJoinGameContainer({
    gameConfig,
    players,
  });

  return (
    <Row justify="center" align="middle">
      <Button
        disabled={isJoinDisabled}
        onClick={() => {
          joinRoomClicked();
        }}
      >
        Join Game
      </Button>
      <Progress
        style={{ margin: "0 2vw" }}
        type="circle"
        percent={percent}
        {...{ status: percent === 0 && "exception" }}
      />
      <Button
        disabled={isLeaveDisabled}
        onClick={() => {
          joinRoomClicked();
        }}
      >
        Leave Game
      </Button>
    </Row>
  );
};
