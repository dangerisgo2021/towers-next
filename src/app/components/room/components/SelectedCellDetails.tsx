import React from "react";
import { Row, Col } from "antd";
import { GiStoneTower, GiCastle } from "react-icons/gi";
import { isNil } from "lodash";

export const SelectedCellDetails = ({
         selectedCell,
         selectedController,
         player,
         currentPlayer,
       }) => (
         <Row
           align="middle"
           justify="space-between"
           style={{
             fontSize: "1.5em",
             border: "1vw solid black",
             backgroundColor: isNil(selectedController.owner)
               ? "grey"
               : selectedController.owner
               ? "blue"
               : "red",
           }}
         >
           <Col>
             {player === 0 ? "Red" : "Blue"} Player
             {player === currentPlayer ? " Go" : " Wait"}
           </Col>
           <Col>
             <Row align="middle" justify="space-around">
               <Col>Selected</Col>
               <Col>
                 <Row align="middle" justify="space-around">
                   <Col>
                     <Row align="middle">
                       <GiStoneTower />
                       <div>{selectedCell.size}</div>
                       <GiCastle
                         style={{
                           color: selectedCell.isCastle ? "gold" : undefined,
                         }}
                       />
                     </Row>
                   </Col>
                 </Row>
               </Col>
             </Row>
           </Col>
         </Row>
       );
