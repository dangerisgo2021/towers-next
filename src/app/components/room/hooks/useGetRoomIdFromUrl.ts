import { useRouter } from "next/router";

export const useGetRoomIdFromUrl = () => {
  const { query = {} } = useRouter();
  const { roomId } = query;
  return roomId;
};
