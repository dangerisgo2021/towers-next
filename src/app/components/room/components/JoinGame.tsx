import React from "react";
import { Button, Row, Progress } from "antd";
import { useJoinGameContainer } from "app/components/room/hooks/useJoinGameContainer";

const defaultGameConfig = { minPlayers: 2 };

export const JoinGame = ({ gameConfig = defaultGameConfig, players = undefined }) => {
  const {
    joinRoomClicked,
    leaveRoomClicked,
    percent,
    isJoinDisabled,
    isLeaveDisabled,
  } = useJoinGameContainer({
    gameConfig,
    players,
  });

  return (
    <Row justify="center" align="middle">
      <Button
        disabled={isJoinDisabled}
        onClick={() => {
          joinRoomClicked();
        }}
      >
        Join Game
      </Button>
      <Progress style={{ margin: "0 2vw" }} type="circle" percent={percent} />
      <Button
        disabled={isLeaveDisabled}
        onClick={() => {
          leaveRoomClicked();
        }}
      >
        Leave Game
      </Button>
    </Row>
  );
};
