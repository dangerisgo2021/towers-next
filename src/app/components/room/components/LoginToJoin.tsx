import React from "react";
import { Button, Row } from "antd";
import { useLoginToJoinContainer } from "app/components/room/hooks/useLoginToJoinContianer";

export const LoginToJoin = () => {
  const { onLoginToJoinCButtonClicked } = useLoginToJoinContainer();
  return (
    <Row justify="space-around">
      <Button block onClick={onLoginToJoinCButtonClicked}>
        Login to Play
      </Button>
    </Row>
  );
};
