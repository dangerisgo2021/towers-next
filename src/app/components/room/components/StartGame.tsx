import React from "react";
import { Button, Row } from "antd";

import { useStartGameContainer } from "app/components/room/hooks/useStartGameContainer";

const defaultGameConfig = { minPlayers: 2 };

export const StartGame = ({ gameConfig = defaultGameConfig, players }) => {
  const { onStartButtonClicked, isDisabled } = useStartGameContainer({
    gameConfig,
    players,
  });

  return (
    <Row justify="center">
      <Button
        type="primary"
        block
        onClick={onStartButtonClicked}
        disabled={isDisabled}
      >
        Start Game Session
      </Button>
    </Row>
  );
};
