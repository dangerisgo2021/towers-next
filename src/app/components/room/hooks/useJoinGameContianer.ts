import { get } from "lodash";
import { useDispatch, useSelector } from "react-redux";
import { joinRoomClicked, leaveRoomClicked } from "state/redux/room/actions";
import { useGetRoomIdFromUrl } from "app/components/room/hooks/useGetRoomIdFromUrl";
import { getSessionUserId } from "state/redux/auth/selectors/getSessionUserId";

export const useJoinGameContainer = ({ gameConfig, players }) => {
  const dispatch = useDispatch();
  const roomId = useGetRoomIdFromUrl();
  const userId = useSelector(getSessionUserId);
  const joined =
    players &&
    players.find((player) => get(player, "profile.userId") === userId);
  const currentPlayerCount = Array.isArray(players) ? players.length : 0;
  const { minPlayers, maxPlayers } = gameConfig;
  const percent = (currentPlayerCount / minPlayers) * 100;
  const isJoinDisabled = joined || currentPlayerCount >= maxPlayers;
  const isLeaveDisabled = !joined;

  return {
    percent,
    isJoinDisabled,
    isLeaveDisabled,
    joinRoomClicked: () => {
      dispatch(joinRoomClicked({ roomId }));
    },
    leaveRoomClicked: () => {
      dispatch(leaveRoomClicked({ roomId }));
    },
  };
};
