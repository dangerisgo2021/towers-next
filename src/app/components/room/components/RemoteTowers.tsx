import React from "react";
import { notification } from "antd";
import { isEqual, isNil } from "lodash";
import { SelectedCellDetails } from "app/components/room/components/SelectedCellDetails";
import { Button } from "antd";
import { VictoryProgress } from "app/components/room/components/VictoryProgress";
import { useRemoteTowersContainer } from "app/components/room/hooks/useRemoteTowersContainer";
import classnames from "classnames";

export const RemoteTowers = ({
  match,
  moveNames,
  currentPlayer,
  player,
  victoryProgress,
}) => {
  const {
    board: { cells, width, height },
    maxTowerSize,
    moves,
    selectedCell,
    setSelectedCell,
  } = useRemoteTowersContainer({ match, moveNames, victoryProgress, player });
  const cellWidth = "50px";

  const selectedController = selectedCell.towerPieces.find((piece, i, array) =>
    array[i + 1] ? array[i + 1].type === "EMPTY" : true
  );
  return (
    <div className="RemoteTowers TowersBoard">
      <SelectedCellDetails
        {...{
          selectedCell,
          selectedController,
          player,
          currentPlayer,
        }}
      />
      <div className="board">
        {cells.map((cell) => (
          <div
            key={cell.id}
            className={classnames("cell", {
              selected: isEqual(cell.location, selectedCell.location),
            })}
            onClick={(e) => {
              e.preventDefault();
              setSelectedCell({ selectedCell: cell });
            }}
          >
            <div className="pieces">
              {cell.towerPieces
                .slice(0) // shallow copy so it doesnt flip flop each render
                .reverse()
                .map((piece) => (
                  <div
                    key={piece.id}
                    className={classnames(
                      "piece",
                      { preview: piece.preview },
                      isNil(piece.owner)
                        ? piece.type
                        : `${piece.type}${piece.owner}`
                    )}
                  />
                ))}
            </div>
          </div>
        ))}
      </div>
      <div className="controls">
        <VictoryProgress
          victoryProgress={victoryProgress}
          currentPlayer={currentPlayer}
          player={player}
        />
        <div className="actions">
          {moves.map((move) => {
            return (
              <Button
                key={move.id}
                style={{
                  gridArea: move.name,
                  gridRow: move.name === "BUILD" ? "1 / span 2" : undefined,
                  height: move.name === "BUILD" ? "100%" : undefined,
                  backgroundColor: player === 0 ? "#deb0b0" : "lightblue",
                  border: "1px solid black",
                }}
                className={classnames("action", move.name)}
                onClick={(e) => {
                  e.preventDefault();
                  if (selectedCell.isCastle && move.name === "BUILD") {
                    notification.open({
                      message: "Cannot build on caste",
                    });
                  } else if (
                    move.name === "BUILD" &&
                    selectedCell.size === maxTowerSize
                  ) {
                    notification.open({
                      message: "Cannot perform build past max tower height",
                    });
                  } else if (selectedController.owner !== currentPlayer) {
                    notification.open({
                      message:
                        "Cannot perform action without controlling square",
                    });
                  } else if (player !== selectedController.owner) {
                    notification.open({
                      message:
                        "Cannot perform action when the tower is not yours",
                    });
                  } else if (player !== currentPlayer) {
                    notification.open({
                      message: "Cannot perform action when its not your turn",
                    });
                  } else if (
                    !isNil(victoryProgress && victoryProgress.winner)
                  ) {
                    notification.open({
                      message: "Cannot perform action when game has a winner",
                    });
                  } else {
                    move.execute(currentPlayer, selectedCell);
                  }
                }}
              >
                {move.text}
              </Button>
            );
          })}
        </div>
      </div>

      <style jsx>{`
        .board {
          display: grid;
          grid-gap: 4px;
          grid-template-columns: repeat(${width}, ${cellWidth});
          grid-template-rows: repeat(${height}, ${cellWidth});
          justify-content: center;
          background-color: dimgray;
          padding: 1vw;
        }

        .piece {
          border: 1px solid black;
        }

        .pieces {
          display: grid;
          grid-template-rows: repeat(${maxTowerSize}, 1fr);
        }

        .cell {
          display: grid;
          grid-template-rows: 1fr min-content;
        }

        .position {
          text-align: center;
        }

        .EMPTY {
          background-color: lightgrey;
        }

        .CASTLE {
          background-color: green;
        }

        .PLAYER0 {
          background-color: red;
        }
        .PLAYER1 {
          background-color: blue;
        }
        .controls {
          border: 1px solid black;
        }
        .test {
          background-color: magenta;
        }
        .action {
          border: 2px solid black;
          background-color: green;
          text-align: center;
          padding: 0;
        }
        .actions {
          display: grid;
          grid-template-areas:
            "BUILD PUSH_LEFT PUSH_RIGHT"
            "BUILD PUSH_UP PUSH_DOWN";
        }

        .info {
          display: grid;
          grid-auto-flow: column;
          grid-template-rows: 1fr 1fr;
        }
        .selected {
          border: 2px solid gold;
        }
        .preview {
          border: 1px solid goldenrod;
        }
        .info-bit {
          border: 1px solid black;
          padding: 3px;
        }
      `}</style>
    </div>
  );
};
