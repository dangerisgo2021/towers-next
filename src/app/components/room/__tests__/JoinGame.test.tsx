import { useJoinGameContainer } from "app/components/room/hooks/useJoinGameContainer";
import { JoinGame } from "app/components/room/components/JoinGame";
import renderer from "react-test-renderer";

jest.mock("app/components/room/hooks/useJoinGameContainer", () => {
  return {
    useJoinGameContainer: jest.fn(),
  };
});

test("JoinGame renders with no props", () => {
  // @ts-ignore useJoinGameContainer is mocked but typescript doesnt know that
  useJoinGameContainer.mockImplementation(() => ({}));
  const component = renderer.create(<JoinGame />).toJSON();
  expect(component).toMatchSnapshot();
});
