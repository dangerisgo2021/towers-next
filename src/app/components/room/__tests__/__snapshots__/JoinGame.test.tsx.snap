// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JoinGame renders with no props 1`] = `
<div
  className="ant-row ant-row-center ant-row-middle"
  style={Object {}}
>
  <button
    className="ant-btn"
    onClick={[Function]}
    type="button"
  >
    <span>
      Join Game
    </span>
  </button>
  <div
    className="ant-progress ant-progress-circle ant-progress-status-normal ant-progress-show-info ant-progress-default"
    style={
      Object {
        "margin": "0 2vw",
      }
    }
  >
    <div
      className="ant-progress-inner"
      style={
        Object {
          "fontSize": 24,
          "height": 120,
          "width": 120,
        }
      }
    >
      <svg
        className="ant-progress-circle"
        style={Object {}}
        viewBox="0 0 100 100"
      >
        <path
          className="ant-progress-circle-trail"
          d="M 50,50 m 0,-47
   a 47,47 0 1 1 0,94
   a 47,47 0 1 1 0,-94"
          fillOpacity="0"
          stroke={null}
          strokeLinecap="round"
          strokeWidth={6}
          style={
            Object {
              "stroke": undefined,
              "strokeDasharray": "295.3097094374406px 295.3097094374406px",
              "strokeDashoffset": "-0px",
              "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s",
            }
          }
        />
        <path
          className="ant-progress-circle-path"
          d="M 50,50 m 0,-47
   a 47,47 0 1 1 0,94
   a 47,47 0 1 1 0,-94"
          fillOpacity="0"
          opacity={0}
          stroke=""
          strokeLinecap="round"
          strokeWidth={6}
          style={
            Object {
              "stroke": undefined,
              "strokeDasharray": "0px 295.3097094374406px",
              "strokeDashoffset": "-0px",
              "transition": "stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s",
            }
          }
        />
      </svg>
      <span
        className="ant-progress-text"
        title="0%"
      >
        0%
      </span>
    </div>
  </div>
  <button
    className="ant-btn"
    onClick={[Function]}
    type="button"
  >
    <span>
      Leave Game
    </span>
  </button>
</div>
`;
